File: curl/curl__.py
#!/usr/bin/env python3
"""
http_curlish.py — curl ⇄ requests power client (2025).. (beta)..

A unified, production-grade Python tool combining:
  • curl-like semantics
  • safe requests.Session patterns
  • streaming uploads
  • retry logic
  • metrics / timing
  • incident response replay

This mirrors *curl behavior*, not libcurl internals.
"""

from __future__ import annotations

import argparse
import json
import logging
import os
import sys
import time
from typing import Any, Dict, Iterable, Optional, Tuple

import requests
from requests.auth import HTTPBasicAuth
from requests.adapters import HTTPAdapter, Retry
from requests_toolbelt.streaming_iterator import StreamingIterator

# =============================================================================
# Logging
# =============================================================================

logging.basicConfig(
    level=logging.INFO,
    format="%(asctime)s [%(levelname)s] %(message)s",
    datefmt="%H:%M:%S",
)

# =============================================================================
# Session Factory (Safe Defaults)
# =============================================================================

def make_session(
    retries: int = 3,
    backoff: float = 0.2,
    timeout: int = 10,
) -> requests.Session:
    """
    Create a hardened requests.Session with retries and timeouts.
    Used by all helpers and CLI paths.
    """
    session = requests.Session()
    retry_strategy = Retry(
        total=retries,
        backoff_factor=backoff,
        status_forcelist=[429, 500, 502, 503, 504],
        allowed_methods=["GET", "POST", "PUT", "DELETE", "PATCH", "OPTIONS"],
    )
    adapter = HTTPAdapter(max_retries=retry_strategy)
    session.mount("https://", adapter)
    session.mount("http://", adapter)
    session.request_timeout = timeout
    return session

# =============================================================================
# Helpers
# =============================================================================

def parse_headers(header_list: list[str]) -> Dict[str, str]:
    headers: Dict[str, str] = {}
    for h in header_list:
        if ":" not in h:
            raise ValueError(f"Invalid header format: {h}")
        k, v = h.split(":", 1)
        headers[k.strip()] = v.strip()
    return headers


def parse_cookies(cookie_string: Optional[str]) -> Optional[Dict[str, str]]:
    if not cookie_string:
        return None
    cookies: Dict[str, str] = {}
    for part in cookie_string.split(";"):
        k, v = part.strip().split("=", 1)
        cookies[k] = v
    return cookies


def load_data_arg(data_arg: str) -> bytes:
    """Load -d / --data argument (literal or @file)."""
    if data_arg.startswith("@"):
        with open(data_arg[1:], "rb") as f:
            return f.read()
    return data_arg.encode()


def stream_fd(fd, chunk_size: int = 8192) -> Iterable[bytes]:
    """
    Stream from file descriptor without buffering entire contents.
    Equivalent to curl -T (unknown-size safe).
    """
    while True:
        chunk = fd.read(chunk_size)
        if not chunk:
            break
        yield chunk

# =============================================================================
# Metrics (curl -w style)
# =============================================================================

class CurlMetrics:
    """Collect curl-like transfer metrics."""

    def __init__(self) -> None:
        self.start = time.perf_counter()
        self.end: Optional[float] = None
        self.bytes_sent = 0
        self.bytes_recv = 0

    def finalize(self) -> None:
        self.end = time.perf_counter()

    def render(self, resp: requests.Response) -> Dict[str, Any]:
        return {
            "http_code": resp.status_code,
            "url_effective": resp.url,
            "content_type": resp.headers.get("Content-Type"),
            "size_upload": self.bytes_sent,
            "size_download": self.bytes_recv,
            "time_total": round((self.end or 0) - self.start, 6),
            "num_redirects": len(resp.history),
            "http_version": getattr(resp.raw, "version", None),
        }

# =============================================================================
# Request Construction
# =============================================================================

def build_request_body(args) -> Tuple[Optional[Any], Optional[Dict[str, Any]]]:
    """
    Build request body using curl semantics:
      -d  buffered data
      -F  multipart form
      -T  streaming upload
    """
    data = None
    files = None

    if args.upload_file:
        fd = open(args.upload_file, "rb")
        try:
            size = os.path.getsize(args.upload_file)
        except OSError:
            size = None
        data = StreamingIterator(size=size, iterator=stream_fd(fd))

    elif args.data:
        data = load_data_arg(args.data)

    elif args.form:
        files = {}
        for entry in args.form:
            k, v = entry.split("=", 1)
            if v.startswith("@"):
                files[k] = open(v[1:], "rb")
            else:
                files[k] = v

    return data, files

# =============================================================================
# High-Level Request Executor
# =============================================================================

def perform_request(args) -> Tuple[requests.Response, bytes, CurlMetrics]:
    session = make_session(timeout=args.max_time or 10)
    headers = parse_headers(args.header)
    cookies = parse_cookies(args.cookie)

    auth = None
    if args.user:
        u, p = args.user.split(":", 1)
        auth = HTTPBasicAuth(u, p)

    data, files = build_request_body(args)
    metrics = CurlMetrics()

    resp = session.request(
        method=args.method.upper(),
        url=args.url,
        headers=headers,
        cookies=cookies,
        auth=auth,
        data=data,
        files=files,
        allow_redirects=args.location,
        verify=not args.insecure,
        timeout=(args.connect_timeout, args.max_time),
        stream=True,
    )

    body_chunks = []
    for chunk in resp.iter_content(8192):
        if chunk:
            metrics.bytes_recv += len(chunk)
            body_chunks.append(chunk)

    metrics.finalize()
    return resp, b"".join(body_chunks), metrics

# =============================================================================
# Pretty Output Helpers
# =============================================================================

def print_response(resp: requests.Response, body: bytes, json_mode: bool) -> None:
    print(f"Status: {resp.status_code}")
    print("Headers:")
    for k, v in resp.headers.items():
        print(f"  {k}: {v}")
    print()

    if json_mode:
        try:
            print(json.dumps(json.loads(body), indent=2))
            return
        except Exception:
            pass

    print(body.decode(errors="replace")[:2000])

# =============================================================================
# CLI
# =============================================================================

def parse_args() -> argparse.Namespace:
    p = argparse.ArgumentParser(
        description="http_curlish — curl ⇄ requests power client"
    )

    p.add_argument("url")
    p.add_argument("-X", "--method", default="GET")
    p.add_argument("-H", "--header", action="append", default=[])
    p.add_argument("-d", "--data")
    p.add_argument("-F", "--form", action="append")
    p.add_argument("-T", "--upload-file")
    p.add_argument("-u", "--user")
    p.add_argument("-b", "--cookie")
    p.add_argument("-k", "--insecure", action="store_true")
    p.add_argument("-L", "--location", action="store_true")
    p.add_argument("-m", "--max-time", type=int)
    p.add_argument("--connect-timeout", type=int, default=5)
    p.add_argument("-i", "--include", action="store_true")
    p.add_argument("-s", "--silent", action="store_true")
    p.add_argument("--json", action="store_true")
    p.add_argument("--metrics", action="store_true")

    return p.parse_args()

def main() -> None:
    args = parse_args()
    resp, body, metrics = perform_request(args)

    if args.include:
        for k, v in resp.headers.items():
            print(f"{k}: {v}")
        print()

    if not args.silent:
        print_response(resp, body, args.json)

    if args.metrics:
        print(json.dumps(metrics.render(resp), indent=2))

# =============================================================================
# Entrypoint
# =============================================================================

if __name__ == "__main__":
    try:
        main()
    except KeyboardInterrupt:
        sys.exit("Cancelled by user")

# =============================================================================
# Extended Documentation
# =============================================================================

"""
http_curlish.py — Extended Documentation
=======================================

WHAT THIS IS
------------
A unified curl ⇄ requests power client for:
  • Incident response
  • AppSec testing
  • Red team replay
  • CI/CD debugging
  • API drift detection

It merges:
  - Safe requests.Session patterns
  - curl-style flags and semantics
  - Streaming uploads (-T)
  - curl -w–style metrics

WHAT THIS IS NOT
----------------
• A libcurl replacement
• A TCP tuning tool
• A browser simulator

STREAMING (-T)
--------------
Streaming uploads avoid buffering entire payloads in memory and
are ideal when:
  • Payload size is unknown
  • You want early write behavior
  • Replaying exploit traffic
  • Stress / chaos testing

SECURITY NOTES
--------------
• Avoid --insecure outside testing
• Streaming uploads bypass Content-Length assumptions
• curl-style CORS tests still require browser validation
• This tool is ideal for forensic replay

EXTENSIONS
----------
• Async engine (httpx / aiohttp)
• HAR export
• Automatic CORS analysis
• GraphQL mode
• Fuzz / chaos streaming
• mTLS / SPIFFE support


EXAMPLES & RECIPES
==================

The following examples mirror common and advanced curl usage.
All examples assume:

    python3 http_curlish.py <args>

------------------------------------------------------------------
BASIC REQUESTS
------------------------------------------------------------------

Equivalent to:
  curl https://example.com

    python3 http_curlish.py https://example.com


HEAD request:
  curl -I https://example.com

    python3 http_curlish.py https://example.com -X HEAD


Follow redirects:
  curl -L https://short.url

    python3 http_curlish.py https://short.url -L


------------------------------------------------------------------
HEADERS & AUTHENTICATION
------------------------------------------------------------------

Add headers:
  curl -H "Accept: application/json" https://api.example.com

    python3 http_curlish.py https://api.example.com \
      -H "Accept: application/json"


Basic auth:
  curl -u user:pass https://api.example.com

    python3 http_curlish.py https://api.example.com \
      -u user:pass


Bearer token:
  curl -H "Authorization: Bearer TOKEN" https://api.example.com

    python3 http_curlish.py https://api.example.com \
      -H "Authorization: Bearer TOKEN"


------------------------------------------------------------------
JSON & FORM POSTS
------------------------------------------------------------------

POST JSON:
  curl -X POST -H "Content-Type: application/json" \
       -d '{"foo":"bar"}' https://api.example.com/post

    python3 http_curlish.py https://api.example.com/post \
      -X POST \
      -H "Content-Type: application/json" \
      -d '{"foo":"bar"}' \
      --json


POST JSON from file:
  curl -d @data.json -H "Content-Type: application/json" -X POST URL

    python3 http_curlish.py URL \
      -X POST \
      -H "Content-Type: application/json" \
      -d @data.json


POST form-urlencoded:
  curl -d "a=1&b=2" https://example.com

    python3 http_curlish.py https://example.com \
      -X POST \
      -d "a=1&b=2"


------------------------------------------------------------------
FILE UPLOADS
------------------------------------------------------------------

Multipart upload:
  curl -F "file=@payload.zip" https://example.com/upload

    python3 http_curlish.py https://example.com/upload \
      -X POST \
      -F "file=@payload.zip"


------------------------------------------------------------------
STREAMING UPLOAD (curl -T DIRTY HACK)
------------------------------------------------------------------

Equivalent to:
  curl -X POST -T payload.bin https://example.com

Streams data without buffering entire file in memory.

    python3 http_curlish.py https://example.com \
      -X POST \
      -T payload.bin


This is useful when:
  • Payload size is unknown
  • Streaming exploit traffic
  • Avoiding memory buffering
  • Chaos / stress testing


------------------------------------------------------------------
COOKIES
------------------------------------------------------------------

Send cookies:
  curl -b "session=abc123" https://example.com

    python3 http_curlish.py https://example.com \
      -b "session=abc123"


------------------------------------------------------------------
TLS / CERTIFICATE TESTING
------------------------------------------------------------------

Ignore TLS verification (testing only):
  curl -k https://self-signed.local

    python3 http_curlish.py https://self-signed.local -k


------------------------------------------------------------------
CORS & SECURITY TESTING
------------------------------------------------------------------

Preflight request:
  curl -X OPTIONS https://api.example.com \
       -H "Origin: https://evil.com" \
       -H "Access-Control-Request-Method: POST"

    python3 http_curlish.py https://api.example.com \
      -X OPTIONS \
      -H "Origin: https://evil.com" \
      -H "Access-Control-Request-Method: POST"


Credentialed cross-origin test:
  curl -H "Origin: https://evil.com" -b "session=abc" URL

    python3 http_curlish.py URL \
      -H "Origin: https://evil.com" \
      -b "session=abc"


------------------------------------------------------------------
METRICS (curl -w EQUIVALENT)
------------------------------------------------------------------

Equivalent to:
  curl -w "%{http_code} %{time_total}" https://example.com

    python3 http_curlish.py https://example.com \
      --silent \
      --metrics


Example output:
{
  "http_code": 200,
  "url_effective": "https://example.com",
  "content_type": "text/html",
  "size_upload": 0,
  "size_download": 1256,
  "time_total": 0.423817,
  "num_redirects": 1,
  "http_version": 11
}


------------------------------------------------------------------
INCIDENT RESPONSE RECIPES
------------------------------------------------------------------

Fast health check:
    python3 http_curlish.py https://api.example.com/health \
      --silent \
      --metrics


Replay suspicious request:
    python3 http_curlish.py https://api.example.com/admin \
      -H "Authorization: Bearer SUSPECT_TOKEN" \
      -i


Before / after deploy diff:
    python3 http_curlish.py https://api.example.com \
      --silent \
      --metrics | sha256sum


------------------------------------------------------------------
COMMON PITFALLS
------------------------------------------------------------------

• -d implies POST (same as curl)
• -T streams, -d buffers
• --json affects output parsing only
• This tool does NOT enforce browser CORS rules
• TCP tuning belongs to OS / libcurl, not Python


------------------------------------------------------------------
PHILOSOPHY
------------------------------------------------------------------

Think of http_curlish as:
  • curl's *intent* in Python
  • a reproducible incident-response probe
  • a safe, readable alternative to shell one-liners

END OF EXAMPLES
------------------------------------------------------------------
"""

File: curl/curl__.go
package main

import (
	"bytes"
	"crypto/tls"
	"encoding/json"
	"flag"
	"fmt"
	"io"
	"net/http"
	"os"
	"strings"
	"time"
)

// =============================================================================
// Types
// =============================================================================

type CurlMetrics struct {
	Start      time.Time
	End        time.Time
	BytesSent  int64
	BytesRecv  int64
	HTTPCode   int
	NumRedirs  int
	TimeTotal  float64
	ContentLen int64
}

type Config struct {
	URL            string
	Method         string
	Headers        []string
	Data           string
	Form           []string
	UploadFile     string
	User           string
	Cookie         string
	Insecure       bool
	Location       bool
	MaxTime        int
	ConnTimeout    int
	Include        bool
	Silent         bool
	JSONMode       bool
	Metrics        bool
}

// =============================================================================
// Client Factory
// =============================================================================

func makeClient(cfg *Config) *http.Client {
	timeout := time.Duration(cfg.MaxTime) * time.Second
	if timeout == 0 {
		timeout = 10 * time.Second
	}

	transport := &http.Transport{
		TLSClientConfig: &tls.Config{
			InsecureSkipVerify: cfg.Insecure,
		},
		MaxIdleConns:        100,
		IdleConnTimeout:     90 * time.Second,
		DisableCompression:  false,
		DisableKeepAlives:   false,
	}

	client := &http.Client{
		Timeout:   timeout,
		Transport: transport,
	}

	if !cfg.Location {
		client.CheckRedirect = func(req *http.Request, via []*http.Request) error {
			return http.ErrUseLastResponse
		}
	}

	return client
}

// =============================================================================
// Header Parsing
// =============================================================================

func parseHeaders(headerList []string) (map[string]string, error) {
	headers := make(map[string]string)
	for _, h := range headerList {
		parts := strings.SplitN(h, ":", 2)
		if len(parts) != 2 {
			return nil, fmt.Errorf("invalid header format: %s", h)
		}
		headers[strings.TrimSpace(parts[0])] = strings.TrimSpace(parts[1])
	}
	return headers, nil
}

func parseCookies(cookieStr string) map[string]string {
	if cookieStr == "" {
		return nil
	}
	cookies := make(map[string]string)
	for _, part := range strings.Split(cookieStr, ";") {
		kv := strings.SplitN(strings.TrimSpace(part), "=", 2)
		if len(kv) == 2 {
			cookies[kv[0]] = kv[1]
		}
	}
	return cookies
}

// =============================================================================
// Body Construction
// =============================================================================

func buildRequestBody(cfg *Config) (io.Reader, string, error) {
	// Streaming upload (-T)
	if cfg.UploadFile != "" {
		file, err := os.Open(cfg.UploadFile)
		if err != nil {
			return nil, "", err
		}
		return file, "application/octet-stream", nil
	}

	// Buffered data (-d)
	if cfg.Data != "" {
		var data []byte
		if strings.HasPrefix(cfg.Data, "@") {
			content, err := os.ReadFile(cfg.Data[1:])
			if err != nil {
				return nil, "", err
			}
			data = content
		} else {
			data = []byte(cfg.Data)
		}
		return bytes.NewReader(data), "application/x-www-form-urlencoded", nil
	}

	// Form data (-F) - simplified version
	if len(cfg.Form) > 0 {
		// For production, use multipart/form-data with proper boundaries
		formData := make(map[string]string)
		for _, entry := range cfg.Form {
			kv := strings.SplitN(entry, "=", 2)
			if len(kv) == 2 {
				formData[kv[0]] = kv[1]
			}
		}
		jsonData, _ := json.Marshal(formData)
		return bytes.NewReader(jsonData), "application/json", nil
	}

	return nil, "", nil
}

// =============================================================================
// Request Execution
// =============================================================================

func performRequest(cfg *Config) (*http.Response, []byte, *CurlMetrics, error) {
	client := makeClient(cfg)
	metrics := &CurlMetrics{Start: time.Now()}

	// Build body
	body, contentType, err := buildRequestBody(cfg)
	if err != nil {
		return nil, nil, nil, err
	}

	// Create request
	req, err := http.NewRequest(cfg.Method, cfg.URL, body)
	if err != nil {
		return nil, nil, nil, err
	}

	// Set headers
	headers, err := parseHeaders(cfg.Headers)
	if err != nil {
		return nil, nil, nil, err
	}
	for k, v := range headers {
		req.Header.Set(k, v)
	}

	// Set content type if we have a body
	if body != nil && contentType != "" {
		req.Header.Set("Content-Type", contentType)
	}

	// Set cookies
	if cfg.Cookie != "" {
		cookies := parseCookies(cfg.Cookie)
		for k, v := range cookies {
			req.AddCookie(&http.Cookie{Name: k, Value: v})
		}
	}

	// Basic auth
	if cfg.User != "" {
		parts := strings.SplitN(cfg.User, ":", 2)
		if len(parts) == 2 {
			req.SetBasicAuth(parts[0], parts[1])
		}
	}

	// Execute request
	resp, err := client.Do(req)
	if err != nil {
		return nil, nil, nil, err
	}
	defer resp.Body.Close()

	// Read response body
	respBody, err := io.ReadAll(resp.Body)
	if err != nil {
		return nil, nil, nil, err
	}

	metrics.End = time.Now()
	metrics.HTTPCode = resp.StatusCode
	metrics.BytesRecv = int64(len(respBody))
	metrics.TimeTotal = metrics.End.Sub(metrics.Start).Seconds()
	metrics.ContentLen = resp.ContentLength

	return resp, respBody, metrics, nil
}

// =============================================================================
// Output
// =============================================================================

func printResponse(resp *http.Response, body []byte, jsonMode bool) {
	fmt.Printf("Status: %d\n", resp.StatusCode)
	fmt.Println("Headers:")
	for k, v := range resp.Header {
		fmt.Printf("  %s: %s\n", k, strings.Join(v, ", "))
	}
	fmt.Println()

	if jsonMode {
		var jsonData interface{}
		if err := json.Unmarshal(body, &jsonData); err == nil {
			pretty, _ := json.MarshalIndent(jsonData, "", "  ")
			fmt.Println(string(pretty))
			return
		}
	}

	output := string(body)
	if len(output) > 2000 {
		output = output[:2000]
	}
	fmt.Println(output)
}

func printMetrics(m *CurlMetrics, resp *http.Response) {
	metrics := map[string]interface{}{
		"http_code":     m.HTTPCode,
		"url_effective": resp.Request.URL.String(),
		"content_type":  resp.Header.Get("Content-Type"),
		"size_upload":   m.BytesSent,
		"size_download": m.BytesRecv,
		"time_total":    fmt.Sprintf("%.6f", m.TimeTotal),
		"content_length": m.ContentLen,
	}

	pretty, _ := json.MarshalIndent(metrics, "", "  ")
	fmt.Println(string(pretty))
}

// =============================================================================
// CLI
// =============================================================================

type headerFlags []string

func (h *headerFlags) String() string {
	return strings.Join(*h, ", ")
}

func (h *headerFlags) Set(value string) error {
	*h = append(*h, value)
	return nil
}

type formFlags []string

func (f *formFlags) String() string {
	return strings.Join(*f, ", ")
}

func (f *formFlags) Set(value string) error {
	*f = append(*f, value)
	return nil
}

func parseArgs() *Config {
	cfg := &Config{}

	var headers headerFlags
	var form formFlags

	flag.StringVar(&cfg.Method, "X", "GET", "HTTP method")
	flag.Var(&headers, "H", "Headers (repeatable)")
	flag.StringVar(&cfg.Data, "d", "", "Request data")
	flag.Var(&form, "F", "Form data (repeatable)")
	flag.StringVar(&cfg.UploadFile, "T", "", "Upload file (streaming)")
	flag.StringVar(&cfg.User, "u", "", "user:password for basic auth")
	flag.StringVar(&cfg.Cookie, "b", "", "Cookie string")
	flag.BoolVar(&cfg.Insecure, "k", false, "Skip TLS verification")
	flag.BoolVar(&cfg.Location, "L", false, "Follow redirects")
	flag.IntVar(&cfg.MaxTime, "m", 10, "Max time in seconds")
	flag.IntVar(&cfg.ConnTimeout, "connect-timeout", 5, "Connection timeout")
	flag.BoolVar(&cfg.Include, "i", false, "Include headers in output")
	flag.BoolVar(&cfg.Silent, "s", false, "Silent mode")
	flag.BoolVar(&cfg.JSONMode, "json", false, "JSON output mode")
	flag.BoolVar(&cfg.Metrics, "metrics", false, "Show metrics")

	flag.Parse()

	if flag.NArg() < 1 {
		fmt.Fprintln(os.Stderr, "Error: URL required")
		flag.Usage()
		os.Exit(1)
	}

	cfg.URL = flag.Arg(0)
	cfg.Headers = headers
	cfg.Form = form

	// curl semantics: -d implies POST
	if cfg.Data != "" && cfg.Method == "GET" {
		cfg.Method = "POST"
	}

	return cfg
}

// =============================================================================
// Main
// =============================================================================

func main() {
	cfg := parseArgs()

	resp, body, metrics, err := performRequest(cfg)
	if err != nil {
		fmt.Fprintf(os.Stderr, "Error: %v\n", err)
		os.Exit(1)
	}

	if cfg.Include {
		for k, v := range resp.Header {
			fmt.Printf("%s: %s\n", k, strings.Join(v, ", "))
		}
		fmt.Println()
	}

	if !cfg.Silent {
		printResponse(resp, body, cfg.JSONMode)
	}

	if cfg.Metrics {
		printMetrics(metrics, resp)
	}
}
